client-js/app/scripts/app.ts

```diff
@@ -283,74 +277,127 @@ window.addEventListener('WebComponentsReady', () => {
-        saveCreds(manager, storage, creds);
+        const result = await mgr.login({
+          username: username,
+          password: password,
+          manager: manager
+        });
+        const credential: Password = {
+          kind: 'username+password',
+          username: username,
+          password: password
+        };
+        const storage = rememberPassword ? window.localStorage
+                                       : window.sessionStorage;
+        saveCredential(manager, storage, credential);
         app.$.loginDialog.close();
         obligation.resolve();
       } catch (error) {
         app.loginError = error.message || error;
         setTimeout(() => app.$.passwordInput.$.input.focus(), 0);
       }
     }
+
+    const authMethods = await mgr.authMethods({manager: manager});
     app.host = manager;
+    app.allowUsernameLogin = authMethods.indexOf('username+password') >= 0;
+    app.allowOAuthLogin = authMethods.indexOf('oauth_token') >= 0;
     app.$.loginButton.addEventListener('click', (event) => doLogin());
     app.$.loginForm.addEventListener('submit', (event) => {
       event.preventDefault();
       event.stopPropagation();
       doLogin();
     });
+    if (app.allowOAuthLogin) {
+      app.$.oauthButton.addEventListener('click', (event) => {
+        startOAuthLogin(mgr, manager)
+      });
+    }
     app.$.loginDialog.open();
     setTimeout(() => app.$.passwordInput.$.input.focus(), 0);
     return promise;
   }
 
-  /**
-   * Load credentials from the given Storage object (session or local).
-   *
-   * If valid credentials are not found, return null.
-   */
-  function loadCreds(manager: string, storage: Storage): Creds {
-    var key = 'labrad-credentials';
-    if (manager) {
-      key += '.' + manager;
-    }
+  async function startOAuthLogin(mgr: ManagerApi, manager: string) {
+    const rememberMe = app.$.rememberPassword.checked;
     try {
-      var creds = JSON.parse(storage.getItem(key));
-      if (creds.hasOwnProperty('username') &&
-          creds.hasOwnProperty('password')) {
-        return {
-          username: creds.username,
-          password: creds.password
-        };
-      }
-    } catch (e) {}
-    return null;
+      const {clientId, clientSecret} = await mgr.oauthInfo({ manager: manager });
+      const redirectUri = `${location.protocol}//${location.host}${prefix}/oauth2callback`;
+      const params = {
+        'response_type': 'code',
+        'client_id': clientId,
+        'redirect_uri': redirectUri,
+        'scope': 'openid email profile',
+        'state': JSON.stringify({
+          path: location.pathname + location.search + location.hash,
+          manager: manager,
+          client_id: clientId,
+          client_secret: clientSecret,
+          redirect_uri: redirectUri,
+          remember_me: rememberMe
+        })
+      };
+      const baseUrl = 'https://accounts.google.com/o/oauth2/v2/auth'
```

Comment:
This is correct as written, there's no `v4` on the `accounts.google.com/o/oauth2` endpoint. Note that the line below for actually renewing the access code to get an id token is `www.googleapis.com/oauth2`. I guess these are just versioned separately. The documentation of the web oauth flow is [here](https://developers.google.com/identity/protocols/OAuth2UserAgent), although we're doing things slightly differently because we have `response_type=code` so that we can exchange the code for an id token. That is more typical of desktop flows, as described in the "Handling the response and exchanging the code" section [here](https://developers.google.com/identity/protocols/OAuth2InstalledApp).